enhancements/sig-policy/122-expanded-hub-templates-access/README.md

```diff
@@ -0,0 +1,310 @@
+}
+```
+
+When the replicated policy controller with the Governance Policy Propagator encounters a
+`spec.hubTemplateOptions.serviceAccountName` field, it'll perform an operation similar to this to either create the
+`TemplateResolver` if it's the first reconcile for that `serviceAccountName` or to use the preexisting one:
+
+```go
+var templateResolver *templates.TemplateResolver
+
+nsName := types.NamespacedName{
+	Namespace: rootPlc.Namespace, Name: replicatedPlc.Spec.HubTemplateOptions.ServiceAccountName,
+}
+
+resolver := &templateResolverWithCancel{
+	lock: &sync.RWMutex{},
+}
+
+resolver.lock.Lock()
+
+oldResolver, loaded := r.TemplateResolvers.LoadOrStore(nsName, resolver)
+if loaded {
+	resolver.lock.Unlock()
+
+	oldResolverTyped := oldResolver.(*templateResolverWithCancel)
+	oldResolverTyped.lock.RLock()
+
+	// Handle the case where resolver is getting cleaned up while the lock is held.
+	// Just try again in this case.
+	if oldResolverTyped.resolver == nil {
+		oldResolverTyped.lock.RUnlock()
+
+		// TODO: Write code to handle repeating the above code block
+	}
+
+	defer oldResolverTyped.lock.RUnlock()
+
+	templateResolver = oldResolverTyped.resolver
+
+	break
+}
+
+// TODO: Code to get a token and instantiate a TemplateResolver instance for the service account and
+// set the values on `resolver`.
+```
+
+There should be a goroutine that periodically cleans up entries in `TemplateResolvers` that are for service accounts no
+longer referenced by any root policy. This is simpler than managing a reference count for each service account.
+
+Another complication is that when the policy's service account name changes, the watches need to be removed for the
+replicated policy on the previous service account `TemplateResolver`. The suggestion is to have another `sync.Map`
+called `policyToServiceAccount` on the `ReplicatedPolicyController` struct to keep track of the service account the
+replicated policy had. The keys are the `ObjectIdentifier` type from the
+[kubernetes-dependency-watches](https://github.com/stolostron/kubernetes-dependency-watches) library representing the
+replicated policy. The values are `types.NamespacedName` values of the service account used. Note that for the default
+template resolver, we will use the special name of `<default>` with an empty namespace. In the `Reconcile` method, if
+the service account changes, the `DynamicWatcher` of the old service account is retrieved from `r.TemplateResolvers` and
+`resolver.DynamicWatcher.RemoveWatcher` is called. This should be done regardless of if hub templates are detected since
+the update of the service account name may have removed hub templates as well. Note that this `policyToServiceAccount`
+map can also be used to clean up watches when a replicated policy is deleted since the original replicated policy spec
+is no longer available.
+
+The other complication is the service account must be periodically refreshed before expiration. This can be handled when
```

Comment:
Correct me if I am wrong, Based on my understanding, 100 service account saved, then run 100 goroutine to watch token expire?

Reply:
Yes, if there are 100 different `serviceAccountName` values, then there'd be 100 goroutines. It should be quite cheap since it just essentially sleeps until near the expiration, then it refreshes the token, and sleeps again.

Reply:
Oh so token expiration is 2 hour so during 2 hour, it sleeps